inference-engine/src/gna_plugin/frontend/scale_factor_calc.hpp

```diff
@@ -576,17 +879,38 @@ class ScaleFactorPerLayer<InferenceEngine::ConcatLayer*> {
             auto restarLayerInfo = LayerInfo(restartedLayer);
             if (restarLayerInfo.isActivation()) {
                 // requantize activation by just changing it's output scale factor
-                quantDataForConCatInput->_dst_quant.SetScale(sourceQuantParams->_dst_quant.GetScale());
-            }
-            if (restarLayerInfo.isConst()) {
+                auto newScaleFactor = sourceQuantParams->_dst_quant.GetScale();
+                auto prevLayer = CNNNetPrevLayer(restartedLayer);
+                while (prevLayer != nullptr && LayerInfo(prevLayer).isNonFunctional()) {
+                    prevLayer = CNNNetPrevLayer(prevLayer);
+                }
+
+                if (fakeQuantize && LayerInfo(prevLayer).isWeightableIdentity()) {
+                    auto weightsScales = generateScaleFactors(MIN_SEARCH_WEIGHTS_VAL, MAX_SEARCH_WEIGHTS_VAL,
+                        MAX_SEARCH_WEIGHTS_VAL - MIN_SEARCH_WEIGHTS_VAL);
+
+                    auto prevLayerQuant = InferenceEngine::getInjectedData<QuantizedLayerParams>(*prevLayer);
+                    auto bestWeightsScale = 1.0f;
+                    if (!fp32eq(prevLayerQuant->_src_quant.GetScale(), newScaleFactor)) {
+                        bestWeightsScale = selectBestWeightsScaleFactors(prevLayerQuant->_src_quant.GetScale(),
+                            newScaleFactor, weightsScales, { 1.0f });
+                    }
+                    if (!fp32eq(bestWeightsScale, prevLayerQuant->_weights_quant.GetScale())) {
+                        gnalog() << "[INFO][Concat] Optimizing weights scale factor for '" << prevLayer->name << "' layer. Change from "
+                            << prevLayerQuant->_weights_quant.GetScale() << " to " << bestWeightsScale << "\n";
+
+                        prevLayerQuant->_weights_quant.SetScale(bestWeightsScale);
+                        prevLayerQuant->_dst_quant.SetScale(prevLayerQuant->_weights_quant.GetScale() * prevLayerQuant->_src_quant.GetScale());
+                        result = ScaleFactorUpdateResult(prevLayer.get());
+                        return true;
+                    }
+                }
+
+                quantDataForConCatInput->_dst_quant.SetScale(newScaleFactor);
+            } else if (restarLayerInfo.isConst()) {
                 gnalog() << "... warning const layer will be requantized\n";
                 quantDataForConCatInput->_dst_quant.SetScale(sourceQuantParams->_dst_quant.GetScale());
             }
-            if (restarLayerInfo.isMemory()) {
```

Comment:
Just to understand, what will be a scenario if a memory layer is a concat input and its scale factor isn't equal to the selected one?

Reply:
That case was not valid. You can't requantize only memory read layer, you have to also requantize memory assign. Otherwise you can end up with mismatch where memory read uses different scale factor than memory assign.